pkg/scheduler/actions/allocate/allocate.go

```diff
@@ -371,10 +358,12 @@ func (alloc *Action) selectBestHyperNode(jobStmts map[string]*framework.Statemen
 	return finalStmt, bestHyperNodeName
 }
 
-func (alloc *Action) allocateResourcesForTasks(tasks *util.PriorityQueue, job *api.JobInfo, queue *api.QueueInfo, allNodes []*api.NodeInfo, hyperNode string) *framework.Statement {
+func (alloc *Action) allocateResourcesForTasks(tasks *util.PriorityQueue, job *api.JobInfo, queue *api.QueueInfo, allNodes []*api.NodeInfo, hyperNode string, softMode bool) *framework.Statement {
```

Comment:
Suggent keep the `allocateResourcesForTasks` func as a basic func without topology related logic, and add extra hard or soft mode process in `allocateResourceForTasksWithTopology`.